# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [30]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [31]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [32]:
spaceship.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


In [33]:
spaceship.isnull().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

In [34]:
#Drop Nulls
spaceship.dropna(inplace=True)

In [35]:
# Extract the first letter of the Cabin and replace NaN values with 'Unknown'
spaceship['Cabin'] = spaceship['Cabin'].str[0].fillna('Unknown')

In [36]:
# Drop PassengerId and Name columns from spaceship DataFrame
spaceship = spaceship.drop(['PassengerId', 'Name'], axis=1)

# Display the updated DataFrame
print(spaceship.head())

  HomePlanet CryoSleep Cabin  Destination   Age    VIP  RoomService  \
0     Europa     False     B  TRAPPIST-1e  39.0  False          0.0   
1      Earth     False     F  TRAPPIST-1e  24.0  False        109.0   
2     Europa     False     A  TRAPPIST-1e  58.0   True         43.0   
3     Europa     False     A  TRAPPIST-1e  33.0  False          0.0   
4      Earth     False     F  TRAPPIST-1e  16.0  False        303.0   

   FoodCourt  ShoppingMall     Spa  VRDeck  Transported  
0        0.0           0.0     0.0     0.0        False  
1        9.0          25.0   549.0    44.0         True  
2     3576.0           0.0  6715.0    49.0        False  
3     1283.0         371.0  3329.0   193.0        False  
4       70.0         151.0   565.0     2.0         True  


In [37]:
# Create dummy variables for non-numerical columns
spaceship = pd.get_dummies(spaceship, columns=spaceship.select_dtypes(include=['object']).columns, drop_first=True)

# Display the updated DataFrame
print(spaceship.head())

    Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0  39.0          0.0        0.0           0.0     0.0     0.0        False   
1  24.0        109.0        9.0          25.0   549.0    44.0         True   
2  58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3  33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4  16.0        303.0       70.0         151.0   565.0     2.0         True   

   HomePlanet_Europa  HomePlanet_Mars  CryoSleep_True  Cabin_B  Cabin_C  \
0               True            False           False     True    False   
1              False            False           False    False    False   
2               True            False           False    False    False   
3               True            False           False    False    False   
4              False            False           False    False    False   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_PSO J318.5-22  \
0  

**Perform Train Test Split**

In [38]:
from sklearn.model_selection import train_test_split

# Assuming 'Transported' is the target variable
X = spaceship.drop('Transported', axis=1)  # Features
y = spaceship['Transported']  # Target

# Perform the train-test split (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Check the shapes of the split data
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)


X_train shape: (5284, 19)
X_test shape: (1322, 19)
y_train shape: (5284,)
y_test shape: (1322,)


**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

- Bagging and Pasting

In [39]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Initialize the model using Bagging with Decision Trees as the base estimator
bagging_model = BaggingClassifier(DecisionTreeClassifier(), n_estimators=100, random_state=42, bootstrap=True)

# Fit the model
bagging_model.fit(X_train, y_train)

# Make predictions
y_pred_bagging = bagging_model.predict(X_test)

# Evaluate the model
bagging_accuracy = accuracy_score(y_test, y_pred_bagging)
print(f'Bagging Accuracy: {bagging_accuracy:.4f}')


Bagging Accuracy: 0.8026


- Random Forests

In [40]:
# Initialize the model using Pasting (bootstrap=False for sampling without replacement)
pasting_model = BaggingClassifier(DecisionTreeClassifier(), n_estimators=100, random_state=42, bootstrap=False)

# Fit the model
pasting_model.fit(X_train, y_train)

# Make predictions
y_pred_pasting = pasting_model.predict(X_test)

# Evaluate the model
pasting_accuracy = accuracy_score(y_test, y_pred_pasting)
print(f'Pasting Accuracy: {pasting_accuracy:.4f}')


Pasting Accuracy: 0.7526


- Gradient Boosting

In [41]:
print("Columns with NaN values:")
print(X.isnull().sum()[X.isnull().sum() > 0])


Columns with NaN values:
Series([], dtype: int64)


In [42]:
from sklearn.experimental import enable_hist_gradient_boosting  # noqa
from sklearn.ensemble import HistGradientBoostingClassifier

# Initialize and fit the HistGradient Boosting model
hgb_model = HistGradientBoostingClassifier(random_state=42)
hgb_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred_hgb = hgb_model.predict(X_test)

# Calculate the accuracy of the model
hgb_accuracy = accuracy_score(y_test, y_pred_hgb)
print(f'HistGradient Boosting Model Accuracy: {hgb_accuracy:.4f}')


HistGradient Boosting Model Accuracy: 0.8185


- Adaptive Boosting

In [43]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Assuming X_train, X_test, y_train, y_test are already defined and processed

# Initialize the AdaBoost model using the default base estimator (DecisionTreeClassifier)
ada_model = AdaBoostClassifier(n_estimators=100, random_state=42)

# Fit the AdaBoost model to the training data
ada_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred_ada = ada_model.predict(X_test)

# Calculate the accuracy of the model
ada_accuracy = accuracy_score(y_test, y_pred_ada)
print(f'AdaBoost Model Accuracy: {ada_accuracy:.4f}')


c:\Users\Usuario\anaconda3\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


AdaBoost Model Accuracy: 0.7958


Which model is the best and why?

AdaBoost Model Accuracy: 0.7958
Bagging Accuracy: 0.8026
HistGradient Boosting Model Accuracy: 0.8185
Pasting Accuracy: 0.7526

Best Model
HistGradient Boosting Model Accuracy: 0.8185 is the best-performing model among the ones tested as it achieved the highest accuracy
